### ISSUE TITLE
Applying sensobol to complex models that do not always converge

### ISSUE THREAD

**Issue by RobinM92** [arnaldpuy/sensobol#9]
Hi @arnaldpuy,
Thank you so much for this package, it is very useful! I am currently applying it on somewhat complex models which can crash or spit out NaNs with certain combinations of parameters. Of course it is also interesting to know about this behavior, but at the moment the sensitivity functions from sensobol expect perfectly behaved model outputs (e.g. length(y) = N*(k+2)).

Is there already a solution for this issue (as I think this was raised before, I also commented in the discussion there. At the moment I am circumventing this by filtering the model response for NaN/NAs and reducing the parameter matrix accordingly, but this can still cause issues as, because of the repetitions, N and Y are then no longer integer multiples of each other. So a better solution (e.g. the option to skip NAs in the SI calculation) would be amazing.
So in short, is there already a workflow to apply this on complex models which might not converge for each set of parameters?

**Comment by arnaldpuy** [comment]
Hi @RobinM92,

Many thanks for using sensobol!

I apologize for not having taken care of this issue yet. Unfortunately the function sobol_indices does not accept NA or NaN values in the model output at this point in time, but I plan to iron this thing out within the next few months (hopefully soon enough). Keep an eye out for new releases.

Sorry about this and all best,

Arnald

**Comment by RobinM92** [comment]
Hi @arnaldpuy,

Thank you so much for replying this quickly (on a Saturday!). At the moment I use the reduced model out put as an input and adapt the parameter matrix by creating a filtered one based on the successful runs. The only catch there is that I need to set N to the amount of successful runs divided by the amount of parameters estimated + 2 (k+2), at least when using eg the saltelli method. Do you think that the calculated sensitivity indices based on this approximation can still be trusted (for now)?

**Comment by arnaldpuy** [comment]
Hi @RobinM92,

Managing NA or NaNs is currently tricky in sensobol because the computation of S_i and T_i for i=1,2,...,k parameters requires to pair the model output from each row of the A, B... matrix with the model output from the correspondent row in the AB_i matrix, and calculate the conditional variance (to know more about what an A, B, AB_i... matrix refers to, have a look at this [paper](https://www.jstatsoft.org/article/view/v102i05). I plan to find a way around this issue in future releases.

Until then: to evaluate whether the computed S_i and T_i values make sense when the model output Y has NA/NaNs, I would first calculate how many NA/NaNs exist in the output, and see where they are located (in the output from the A, AB_i matrix, etc). If there are many of them, they will probably bias all sensitivity indices. If there are just a few, I would have a look and try to see whether they relate to a specific parameter (e.g., if they relate to parameter x_i, you will find NA / NaN values mostly in the output from the AB_i matrix). Under these circumstances you may be able to trust the sensitivity indices for the other parameters. If they are spread all over the model output, the indices may not be reliable.

Hope this helps,

Arnald
